### ISSUE TITLE
Proposal: plugin mechanism for Dragonfly

### ISSUE THREAD

**Issue by lowzj** [dragonflyoss/Dragonfly#1309]
## Why you need it?

More and more customized requirements that are not covered by Dragonfly come out. 
* support different protocols(now only `HTTP/HTTPS`): https://github.com/dragonflyoss/Dragonfly/issues/1119#issuecomment-561096976
* different scheduler: https://github.com/dragonflyoss/Dragonfly/issues/1002
* different ways to get supernode list: https://github.com/dragonflyoss/Dragonfly/issues/1293
* add a new distribute pattern based on p2p: https://github.com/dragonflyoss/Dragonfly/issues/1284
* security requirements: https://github.com/dragonflyoss/Dragonfly/issues/1204

and so on.

Here is a simple plugin implementation in `supernode`: [supernode/plugins](https://github.com/dragonflyoss/Dragonfly/tree/master/supernode/plugins). 
But I think that it's not enough for the whole Dragonfly project including components`supernode`/`dfget`/`dfdaemon`. So there should be a plugin mechanism for Dragonfly, and put its implementation into `pkg` package.

## How it could be?

Here is a [document](https://docs.google.com/document/d/1oFX8r9eRgcFs67BxDkKtZUfWG1_4ASXBvv6oQL_pAlA/edit#heading=h.y84g93qkbeq9) comparing the 3 ways to implement plugin:
* vendor the core code
* golang-plugins
* hashicorp/go-plugin: https://github.com/hashicorp/go-plugin

Dragonfly can select a suitable solution to implement as its foundation for users to implements their own plugins conveniently.

Based on this, Dragonfly abstracts some pluggable components in the next step.
Furthermore, a new individual project `dragonflyoss/plugins` could be created for users to submit their plugin implementation which cannot be merged into the main project for the time being but hope to be developed in the community.

**Comment by inoc603** [comment]
I think before we choose a plugin mechanism, we should refactor our code to identify the parts where customization is necessary, and make them configurable in code first.

For example, if we need adding customizable protocols in dfdaemon, we should first define the interface for a custom protocol, and probably implement http/https as one. 

```go
type Protocol interface {
  // ...
}

func WithProtocol(p Protocol) Option {
  // ...
}
```

Then we can define the plugin interface for user to customize. Take go plugin as an example, we may ask the user to provide a specific function in their plugin

```go
func ProtocolFactory(config map[string]interface{}) Protocol
```

Then for plugins that would be widely used, we can implement them in dragonfly. For other more specific requirements, the user may choose to implement them as plugin or fork the project, and both method will use the same interface in code.

**Comment by lowzj** [comment]
> I think before we choose a plugin mechanism, we should refactor our code to identify the parts where customization is necessary, and make them configurable in code first.

@inoc603 agreed. It's necessary to define an interface for a component before we make it pluggable and configurable. 

But I think that the definition interface of pluggable components and the common implementation of plugin mechanism can be executed in parallel.

The plugin mechanism should provide these features:
* provide a way for Dragonfly to select a specific implementation of a plugin
* provide a way for user to register their own implementation of a plugin
* provide a way to construct a plugin with configuration

And Dragonfly components can use this mechanism to make themself pluggable in a unified way.

**Comment by lowzj** [comment]
I will create a project [here](https://github.com/dragonflyoss/Dragonfly/projects) to list and track tasks about interface definition.

**Comment by inoc603** [comment]
> > I think before we choose a plugin mechanism, we should refactor our code to identify the parts where customization is necessary, and make them configurable in code first.
> 
> @inoc603 agreed. It's necessary to define an interface for a component before we make it pluggable and configurable.
> 
> But I think that the definition interface of pluggable components and the common implementation of plugin mechanism can be executed in parallel.
> 
> The plugin mechanism should provide these features:
> 
> * provide a way for Dragonfly to select a specific implementation of a plugin
> * provide a way for user to register their own implementation of a plugin
> * provide a way to construct a plugin with configuration
> 
> And Dragonfly components can use this mechanism to make themself pluggable in a unified way.

For dynamically loaded plugins, I think we can use a very simple configuration:

```yaml
# a list of plugins that will be dynamically loaded
plugins:
-
  # type of the protocol, should be one of the plugin types provided by dragonfly
  type: protocol
  # name of the plugin
  name: nfs
  # how to load the plugin
  # take go plugin as an example, we need the path of the plugin
  path: /path/to/plugin
  # configuration for the plugin
  args:
    # any valid yaml
```

**Comment by lowzj** [comment]
> For dynamically loaded plugins, I think we can use a very simple configuration:
> 
> ```yaml
> # a list of plugins that will be dynamically loaded
> plugins:
> -
>   # type of the protocol, should be one of the plugin types provided by dragonfly
>   type: protocol
>   # name of the plugin
>   name: nfs
>   # how to load the plugin
>   # take go plugin as an example, we need the path of the plugin
>   path: /path/to/plugin
>   # configuration for the plugin
>   args:
>     # any valid yaml
> ```

Where should the configurations be placed? An individual file or integrated with current `supernod.yml/dfget.yml/dfdaemon.yml`?

**Comment by zcc35357949** [comment]
Is there any further updates for this issue?
